File: frontend/lib/professor.ts
import { deleteJson, getJson, patchJson, postJson } from './apiClient'

export type ProgramTrack = {
  id: number
  niveau: string
  filiere: string
  title: string
  status: string
}

export type CourseOffering = {
  id: number
  subject_id: number
  subject_title: string
  track: ProgramTrack
  professor_user_id: number
  title: string
  status: string
}

export type LiveSession = {
  id: number
  course_offering_id: number
  title: string
  description: string
  starts_at: string
  ends_at: string
  status: 'scheduled' | 'live' | 'completed' | 'cancelled' | string
  join_url: string
  vdocipher_live_id: string
  notification_status: string
  created_at: string
}

export type ProfessorLiveSession = LiveSession & {
  has_stream_credentials: boolean
}

export type LiveSessionStreamCredentials = {
  id: number
  stream_ingest_url: string
  stream_key: string
}

export type StudentLiveSession = LiveSession & {
  offering_title: string
  subject_title: string
  niveau: string
  filiere: string
  teacher_name: string
  viewer_url: string
  can_join: boolean
  provider: 'vdocipher' | string
}

export type LiveSessionEmbed = {
  id: number
  title: string
  status: string
  provider: 'vdocipher' | string
  embed_url: string
  chat_embed_url: string
  vdocipher_live_id: string
}

export type LiveProviderConfig = {
  provider: 'vdocipher' | string
  has_api_secret: boolean
  can_auto_create: boolean
  missing: string[]
  create_endpoint_configured: boolean
}

export type LiveSessionInteraction = {
  id: number
  live_session_id: number
  course_offering_id: number
  professor_user_id: number
  student_user_id: number
  student_name: string
  kind: 'question' | 'message' | string
  body: string
  status: 'pending' | 'answered' | 'hidden' | 'deleted' | string
  answer: string
  answered_by_user_id?: number | null
  answered_at?: string | null
  deleted_at?: string | null
  created_at: string
  updated_at: string
}

export type LiveSessionCheckpoint = {
  id: number
  live_session_id: number
  course_offering_id: number
  professor_user_id: number
  title: string
  prompt: string
  checkpoint_type: 'prompt' | 'quiz' | string
  status: 'active' | 'closed' | 'deleted' | string
  created_at: string
  closed_at?: string | null
}

export type ChangeRequest = {
  id: number
  course_offering_id: number
  target_type: string
  target_id: number
  change_type: string
  proposed_patch_json: Record<string, unknown>
  current_snapshot_json: Record<string, unknown>
  status: string
  admin_note: string
  created_at: string
  reviewed_at?: string | null
}

export type ProfessorDashboard = {
  offerings: CourseOffering[]
  active_offering: CourseOffering | null
  upcoming_live_sessions: ProfessorLiveSession[]
  pending_change_requests: ChangeRequest[]
  chat_unread_count: number
  chat_pinned_count: number
}

export type ChatParticipant = {
  id: number
  full_name: string
  avatar_url: string
  tier: string
}

export type ProfessorConversation = {
  id: number
  course_offering_id: number
  offering_title: string
  subject_title: string
  niveau: string
  filiere: string
  professor: ChatParticipant
  student: ChatParticipant
  status: string
  last_message_preview: string
  unread_for_professor: number
  unread_for_student: number
  is_pinned_by_professor: boolean
  created_at: string
  updated_at: string
  last_message_at: string
}

export type StudentProfessorThread = {
  course_offering_id: number
  offering_title: string
  subject_title: string
  niveau: string
  filiere: string
  professor: ChatParticipant
  conversation: ProfessorConversation | null
  last_message_preview: string
  last_message_sender_role: string
  unread_count: number
  last_message_at?: string | null
}

export type ProfessorMessage = {
  id: number
  conversation_id: number
  sender_user_id: number
  sender_role: string
  body: string
  attachment_url: string
  attachment_mime_type: string
  attachment_name: string
  attachment_size: number
  status: string
  created_at: string
  read_at?: string | null
}

export type StudentProfessorChatStatus = {
  eligible: boolean
  reason: string
  offerings: CourseOffering[]
  conversations: ProfessorConversation[]
  teacher_threads?: StudentProfessorThread[]
}

type OffsetPageParams = { limit?: number; offset?: number }
type CursorPageParams = { limit?: number; before_id?: number }
type LiveInteractionPageParams = CursorPageParams & { status?: string; kind?: string }
type StudentLiveInteractionPageParams = CursorPageParams & { kind?: string }

export async function getProfessorDashboard() {
  return getJson<ProfessorDashboard>('/professor/dashboard')
}

export async function listProfessorOfferings() {
  return getJson<CourseOffering[]>('/professor/offerings')
}

export async function listProfessorLiveSessions() {
  return getJson<ProfessorLiveSession[]>('/professor/live-sessions')
}

export async function getProfessorLiveProviderConfig() {
  return getJson<LiveProviderConfig>('/professor/live-provider-config')
}

export type LiveSessionInput = {
  course_offering_id: number
  title: string
  description: string
  starts_at: string
  ends_at: string
  join_url: string
  vdocipher_live_id?: string
  stream_ingest_url?: string
  stream_key?: string
  auto_create_vdocipher?: boolean
  chat_mode?: string
}

export async function createProfessorLiveSession(payload: LiveSessionInput) {
  return postJson<ProfessorLiveSession>('/professor/live-sessions', payload)
}

export async function updateProfessorLiveSession(id: number, payload: Partial<LiveSessionInput> & { status?: string }) {
  return patchJson<ProfessorLiveSession>(`/professor/live-sessions/${id}`, payload)
}

export async function notifyProfessorLiveSession(id: number) {
  return postJson<ProfessorLiveSession>(`/professor/live-sessions/${id}/notify`)
}

export async function startProfessorLiveSession(id: number) {
  return postJson<ProfessorLiveSession>(`/professor/live-sessions/${id}/start`)
}

export async function endProfessorLiveSession(id: number) {
  return postJson<ProfessorLiveSession>(`/professor/live-sessions/${id}/end`)
}

export async function cancelProfessorLiveSession(id: number) {
  return postJson<ProfessorLiveSession>(`/professor/live-sessions/${id}/cancel`)
}

export async function deleteProfessorLiveSession(id: number) {
  return deleteJson<{ ok: boolean }>(`/professor/live-sessions/${id}`)
}

export async function getProfessorLiveEmbed(id: number) {
  return getJson<LiveSessionEmbed>(`/professor/live-sessions/${id}/embed`)
}

export async function revealProfessorLiveStreamCredentials(id: number) {
  return postJson<LiveSessionStreamCredentials>(`/professor/live-sessions/${id}/stream-credentials/reveal`)
}

export async function listStudentLiveSessions() {
  return getJson<StudentLiveSession[]>('/professor/student-live-sessions')
}

export async function getStudentLiveEmbed(id: number) {
  return getJson<LiveSessionEmbed>(`/professor/student-live-sessions/${id}/embed`)
}

export async function listProfessorLiveInteractions(id: number, paramsOrStatus: string | LiveInteractionPageParams = {}) {
  const params = typeof paramsOrStatus === 'string'
    ? (paramsOrStatus ? { status: paramsOrStatus } : {})
    : paramsOrStatus
  return getJson<LiveSessionInteraction[]>(`/professor/live-sessions/${id}/interactions`, {
    params: Object.keys(params).length ? params : undefined,
  })
}

export async function answerProfessorLiveInteraction(id: number, answer: string) {
  return patchJson<LiveSessionInteraction>(`/professor/live-sessions/interactions/${id}`, { answer })
}

export async function patchProfessorLiveInteraction(id: number, patch: { status?: string; answer?: string }) {
  return patchJson<LiveSessionInteraction>(`/professor/live-sessions/interactions/${id}`, patch)
}

export async function deleteProfessorLiveInteraction(id: number) {
  return deleteJson<LiveSessionInteraction>(`/professor/live-sessions/interactions/${id}`)
}

export async function listStudentLiveInteractions(id: number, params: StudentLiveInteractionPageParams = {}) {
  return getJson<LiveSessionInteraction[]>(`/professor/student-live-sessions/${id}/interactions`, {
    params: Object.keys(params).length ? params : undefined,
  })
}

export async function createStudentLiveInteraction(id: number, body: string, kind = 'question') {
  return postJson<LiveSessionInteraction>(`/professor/student-live-sessions/${id}/interactions`, { body, kind })
}

export async function listProfessorLiveCheckpoints(id: number) {
  return getJson<LiveSessionCheckpoint[]>(`/professor/live-sessions/${id}/checkpoints`)
}

export async function createProfessorLiveCheckpoint(id: number, payload: { title: string; prompt?: string; checkpoint_type?: string }) {
  return postJson<LiveSessionCheckpoint>(`/professor/live-sessions/${id}/checkpoints`, payload)
}

export async function patchProfessorLiveCheckpoint(id: number, patch: { status?: string }) {
  return patchJson<LiveSessionCheckpoint>(`/professor/live-sessions/checkpoints/${id}`, patch)
}

export async function listStudentLiveCheckpoints(id: number) {
  return getJson<LiveSessionCheckpoint[]>(`/professor/student-live-sessions/${id}/checkpoints`)
}

export async function listProfessorChangeRequests(status = 'pending') {
  return getJson<ChangeRequest[]>('/professor/change-requests', { params: { status } })
}

export async function listProfessorConversations(params: { q?: string; unread?: boolean; pinned?: boolean } & OffsetPageParams = {}) {
  return getJson<ProfessorConversation[]>('/professor/chat/conversations', { params })
}

export async function listProfessorMessages(conversationId: number, params: CursorPageParams = {}) {
  return getJson<ProfessorMessage[]>(`/professor/chat/conversations/${conversationId}/messages`, { params })
}

export async function sendProfessorMessage(conversationId: number, body: string) {
  return postJson<ProfessorMessage>(`/professor/chat/conversations/${conversationId}/messages`, { body })
}

export async function sendProfessorImageMessage(conversationId: number, file: File, body = '') {
  const form = new FormData()
  form.append('file', file)
  if (body.trim()) form.append('body', body.trim())
  const data = await postJson<ProfessorMessage>(`/professor/chat/conversations/${conversationId}/images`, form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

export async function updateProfessorChatMessage(messageId: number, body: string) {
  return patchJson<ProfessorMessage>(`/professor/chat/messages/${messageId}`, { body })
}

export async function deleteProfessorChatMessage(messageId: number) {
  await deleteJson(`/professor/chat/messages/${messageId}`)
}

export async function patchProfessorConversation(conversationId: number, patch: { is_pinned_by_professor?: boolean; mark_read?: boolean }) {
  return patchJson<ProfessorConversation>(`/professor/chat/conversations/${conversationId}`, patch)
}

export async function getStudentProfessorChat() {
  return getJson<StudentProfessorChatStatus>('/professor/student-chat')
}

export async function startStudentProfessorConversation(courseOfferingId: number, body: string) {
  return postJson<ProfessorConversation>('/professor/student-chat/conversations', {
    course_offering_id: courseOfferingId,
    body,
  })
}

export async function listStudentProfessorMessages(conversationId: number, params: CursorPageParams = {}) {
  return getJson<ProfessorMessage[]>(`/professor/student-chat/conversations/${conversationId}/messages`, { params })
}

export async function sendStudentProfessorMessage(conversationId: number, body: string) {
  return postJson<ProfessorMessage>(`/professor/student-chat/conversations/${conversationId}/messages`, { body })
}

export async function sendStudentProfessorImageMessage(conversationId: number, file: File, body = '') {
  const form = new FormData()
  form.append('file', file)
  if (body.trim()) form.append('body', body.trim())
  const data = await postJson<ProfessorMessage>(`/professor/student-chat/conversations/${conversationId}/images`, form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

export function chatMediaUrl(value?: string | null) {
  if (!value) return ''
  if (/^(https?:|blob:|data:)/i.test(value)) return value
  return value
}

File: backend/app/schemas/professor.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field, field_validator

from app.schemas.limits import (
    JsonBounds,
    LongText,
    ShortText,
    StrictInputModel,
    TokenText,
    UrlText,
    validate_bounded_json_object,
)


PROFESSOR_CHANGE_REQUEST_JSON_BOUNDS = JsonBounds(
    max_container_depth=6,
    max_dict_items=250,
    max_list_items=250,
    max_string_length=10000,
    max_total_bytes=128 * 1024,
)


class ProgramTrackOut(BaseModel):
    id: int
    niveau: str
    filiere: str
    title: str
    status: str

    model_config = {"from_attributes": True}


class CourseOfferingOut(BaseModel):
    id: int
    subject_id: int
    subject_title: str
    track: ProgramTrackOut
    professor_user_id: int
    title: str
    status: str


class LiveSessionOut(BaseModel):
    id: int
    course_offering_id: int
    title: str
    description: str
    starts_at: datetime
    ends_at: datetime
    status: str
    join_url: str
    vdocipher_live_id: str
    notification_status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class ProfessorLiveSessionOut(LiveSessionOut):
    has_stream_credentials: bool = False


class LiveSessionStreamCredentialsOut(BaseModel):
    id: int
    stream_ingest_url: str = ""
    stream_key: str = ""


class LiveSessionViewerOut(LiveSessionOut):
    offering_title: str = ""
    subject_title: str = ""
    niveau: str = ""
    filiere: str = ""
    teacher_name: str = ""
    viewer_url: str = ""
    can_join: bool = False
    provider: str = "vdocipher"


class LiveSessionEmbedOut(BaseModel):
    id: int
    title: str
    status: str
    provider: str = "vdocipher"
    embed_url: str
    chat_embed_url: str = ""
    vdocipher_live_id: str


class LiveProviderConfigOut(BaseModel):
    provider: str = "vdocipher"
    has_api_secret: bool
    can_auto_create: bool
    missing: list[str] = []
    create_endpoint_configured: bool


class LiveSessionIn(StrictInputModel):
    course_offering_id: int
    title: str = Field(min_length=1, max_length=255)
    description: LongText = ""
    starts_at: datetime
    ends_at: datetime
    join_url: UrlText = ""
    vdocipher_live_id: ShortText = ""
    stream_ingest_url: UrlText = ""
    stream_key: TokenText = ""
    auto_create_vdocipher: bool = False
    chat_mode: ShortText = "off"


class LiveSessionUpdateIn(StrictInputModel):
    course_offering_id: Optional[int] = None
    title: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[LongText] = None
    starts_at: Optional[datetime] = None
    ends_at: Optional[datetime] = None
    join_url: Optional[UrlText] = None
    vdocipher_live_id: Optional[ShortText] = None
    stream_ingest_url: Optional[UrlText] = None
    stream_key: Optional[TokenText] = None
    status: Optional[ShortText] = None


class LiveSessionInteractionOut(BaseModel):
    id: int
    live_session_id: int
    course_offering_id: int
    professor_user_id: int
    student_user_id: int
    student_name: str = ""
    kind: str
    body: str
    status: str
    answer: str = ""
    answered_by_user_id: Optional[int] = None
    answered_at: Optional[datetime] = None
    deleted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class LiveSessionInteractionIn(StrictInputModel):
    kind: ShortText = "question"
    body: str = Field(min_length=1, max_length=2000)


class LiveSessionInteractionPatchIn(StrictInputModel):
    status: Optional[ShortText] = None
    answer: Optional[str] = Field(default=None, max_length=4000)


class LiveSessionCheckpointOut(BaseModel):
    id: int
    live_session_id: int
    course_offering_id: int
    professor_user_id: int
    title: str
    prompt: str = ""
    checkpoint_type: str
    status: str
    created_at: datetime
    closed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class LiveSessionCheckpointIn(StrictInputModel):
    title: str = Field(min_length=1, max_length=255)
    prompt: str = Field(default="", max_length=4000)
    checkpoint_type: ShortText = "prompt"


class LiveSessionCheckpointPatchIn(StrictInputModel):
    status: Optional[ShortText] = None


class ProfessorChangeRequestOut(BaseModel):
    id: int
    course_offering_id: int
    target_type: str
    target_id: int
    change_type: str
    proposed_patch_json: dict[str, Any]
    current_snapshot_json: dict[str, Any]
    status: str
    admin_note: str
    created_at: datetime
    reviewed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ProfessorChangeRequestIn(StrictInputModel):
    course_offering_id: int
    target_type: ShortText
    target_id: int
    change_type: ShortText = "update_fields"
    proposed_patch_json: dict[str, Any]
    current_snapshot_json: dict[str, Any] = Field(default_factory=dict)

    @field_validator("proposed_patch_json", "current_snapshot_json")
    @classmethod
    def validate_change_request_json(cls, value: dict[str, Any]) -> dict[str, Any]:
        return validate_bounded_json_object(value, bounds=PROFESSOR_CHANGE_REQUEST_JSON_BOUNDS)


# ── Professor Studio ────────────────────────────────────────────────────────

STUDIO_OP_TYPES = {"create", "update_fields", "update_content", "delete", "reorder"}
STUDIO_ENTITY_TYPES = {"chapter", "lesson", "tab"}


class StudioTabOut(BaseModel):
    id: int
    label: str
    tab_type: str
    status: str
    order: int
    content: str = ""
    resource_url: str = ""
    renderer_key: str = ""
    config_json: dict[str, Any] = Field(default_factory=dict)


class StudioLessonOut(BaseModel):
    id: int
    title: str
    description: str = ""
    item_type: str
    status: str
    order: int
    is_free_preview: bool = False
    required_tier: str = ""
    duration_seconds: int = 0
    # Primary VdoCipher video id for the lesson (the player resolves the stream
    # from this). Empty when the lesson has no hosted video.
    video_id: str = ""
    tabs: list[StudioTabOut] = Field(default_factory=list)


class StudioChapterOut(BaseModel):
    id: int
    title: str
    description: str = ""
    status: str
    order: int
    is_free_preview: bool = False
    required_tier: str = ""
    lessons: list[StudioLessonOut] = Field(default_factory=list)


class StudioTreeOut(BaseModel):
    course_offering_id: int
    offering_title: str
    subject_title: str
    chapters: list[StudioChapterOut] = Field(default_factory=list)
    has_pending_request: bool = False
    pending_request_id: Optional[int] = None
    # Ids of existing items that are referenced by a still-pending change
    # request, so the studio can flag them "En attente de validation".
    pending_chapter_ids: list[int] = Field(default_factory=list)
    pending_lesson_ids: list[int] = Field(default_factory=list)
    pending_tab_ids: list[int] = Field(default_factory=list)


class StudioOperationIn(StrictInputModel):
    op_type: ShortText
    entity_type: ShortText
    # Real id of an existing item (None for create ops).
    target_id: Optional[int] = None
    # Temporary id assigned by the studio for items created in this batch.
    client_ref: ShortText = ""
    # Parent reference: a real id (int as str) or a sibling create op's client_ref.
    parent_ref: ShortText = ""
    payload: dict[str, Any] = Field(default_factory=dict)
    snapshot: dict[str, Any] = Field(default_factory=dict)

    @field_validator("op_type")
    @classmethod
    def validate_op_type(cls, value: str) -> str:
        if value not in STUDIO_OP_TYPES:
            raise ValueError(f"Unsupported op_type: {value}")
        return value

    @field_validator("entity_type")
    @classmethod
    def validate_entity_type(cls, value: str) -> str:
        if value not in STUDIO_ENTITY_TYPES:
            raise ValueError(f"Unsupported entity_type: {value}")
        return value

    @field_validator("payload", "snapshot")
    @classmethod
    def validate_op_json(cls, value: dict[str, Any]) -> dict[str, Any]:
        return validate_bounded_json_object(value, bounds=PROFESSOR_CHANGE_REQUEST_JSON_BOUNDS)


class StudioSubmitIn(StrictInputModel):
    course_offering_id: int
    summary: LongText = ""
    operations: list[StudioOperationIn] = Field(default_factory=list, max_length=200)


class ProfessorChangeOperationOut(BaseModel):
    id: int
    seq: int
    op_type: str
    entity_type: str
    target_id: Optional[int] = None
    client_ref: str = ""
    parent_ref: str = ""
    payload_json: dict[str, Any] = Field(default_factory=dict)
    snapshot_json: dict[str, Any] = Field(default_factory=dict)
    status: str
    applied_target_id: Optional[int] = None
    error_detail: str = ""

    model_config = {"from_attributes": True}


class ProfessorChangeRequestDetailOut(ProfessorChangeRequestOut):
    summary: str = ""
    professor_name: str = ""
    professor_email: str = ""
    offering_title: str = ""
    operations: list[ProfessorChangeOperationOut] = Field(default_factory=list)


class ProfessorChangeRequestSummaryOut(BaseModel):
    id: int
    course_offering_id: int
    offering_title: str = ""
    summary: str = ""
    status: str
    operation_count: int = 0
    pending_count: int = 0
    applied_count: int = 0
    rejected_count: int = 0
    admin_note: str = ""
    created_at: datetime
    reviewed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class AdminChangeRequestListItemOut(BaseModel):
    id: int
    course_offering_id: int
    offering_title: str = ""
    professor_name: str = ""
    professor_email: str = ""
    summary: str = ""
    status: str
    operation_count: int = 0
    pending_count: int = 0
    created_at: datetime
    reviewed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class AdminOperationDecisionIn(StrictInputModel):
    operation_id: int
    decision: ShortText  # approve | reject

    @field_validator("decision")
    @classmethod
    def validate_decision(cls, value: str) -> str:
        if value not in {"approve", "reject"}:
            raise ValueError("decision must be 'approve' or 'reject'")
        return value


class AdminReviewIn(StrictInputModel):
    decisions: list[AdminOperationDecisionIn] = Field(default_factory=list, max_length=200)
    admin_note: LongText = ""


class ProfessorDashboardOut(BaseModel):
    offerings: list[CourseOfferingOut]
    active_offering: Optional[CourseOfferingOut]
    upcoming_live_sessions: list[ProfessorLiveSessionOut]
    pending_change_requests: list[ProfessorChangeRequestSummaryOut]
    chat_unread_count: int
    chat_pinned_count: int


class ChatParticipantOut(BaseModel):
    id: int
    full_name: str
    avatar_url: str = ""
    tier: str = "basic"


class ProfessorChatConversationOut(BaseModel):
    id: int
    course_offering_id: int
    offering_title: str
    subject_title: str
    niveau: str
    filiere: str
    professor: ChatParticipantOut
    student: ChatParticipantOut
    status: str
    last_message_preview: str
    unread_for_professor: int
    unread_for_student: int
    is_pinned_by_professor: bool
    created_at: datetime
    updated_at: datetime
    last_message_at: datetime


class StudentProfessorThreadOut(BaseModel):
    course_offering_id: int
    offering_title: str
    subject_title: str
    niveau: str
    filiere: str
    professor: ChatParticipantOut
    conversation: Optional[ProfessorChatConversationOut] = None
    last_message_preview: str = ""
    last_message_sender_role: str = ""
    unread_count: int = 0
    last_message_at: Optional[datetime] = None


class ProfessorChatMessageOut(BaseModel):
    id: int
    conversation_id: int
    sender_user_id: int
    sender_role: str
    body: str
    attachment_url: str = ""
    attachment_mime_type: str = ""
    attachment_name: str = ""
    attachment_size: int = 0
    status: str
    created_at: datetime
    read_at: Optional[datetime] = None


class ChatMessageIn(StrictInputModel):
    body: str = Field(min_length=1, max_length=4000)


class ChatMessagePatchIn(StrictInputModel):
    body: str = Field(min_length=1, max_length=4000)


class StudentStartConversationIn(ChatMessageIn):
    course_offering_id: int


class ChatConversationPatchIn(StrictInputModel):
    is_pinned_by_professor: Optional[bool] = None
    mark_read: bool = False


class StudentProfessorChatStatusOut(BaseModel):
    eligible: bool
    reason: str = ""
    offerings: list[CourseOfferingOut] = []
    conversations: list[ProfessorChatConversationOut] = []
    teacher_threads: list[StudentProfessorThreadOut] = []
